consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,434 @@
+		nodeAddr: addr,
+		state: state[V, H]{
+			h:           chain.Height(),
+			lockedRound: -1,
+			validRound:  -1,
+		},
+		messages:         newMessages[V, H, A](),
+		futureMessages:   newMessages[V, H, A](),
+		futureMessagesMu: &sync.Mutex{},
+		timeoutPropose:   tmPropose,
+		timeoutPrevote:   tmPrevote,
+		timeoutPrecommit: tmPrecommit,
+		application:      app,
+		blockchain:       chain,
+		validators:       vals,
+		listeners:        listeners,
+		broadcasters:     broadcasters,
+		scheduledTms:     make([]timeout, 0),
+		timeoutsCh:       make(chan timeout),
+		proposalsCh:      make(chan Proposal[V, H, A]),
+		prevotesCh:       make(chan Prevote[H, A]),
+		precommitsCh:     make(chan Precommit[H, A]),
+		quit:             make(chan struct{}),
+	}
+}
+
+func (t *Tendermint[V, H, A]) Start() {
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+
+		t.startRound(0)
+
+		// Todo: check message signature everytime a message is received.
+		// For the time being it can be assumed the signature is correct.
+
+		for {
+			select {
+			case <-t.quit:
+				return
+			case tm := <-t.timeoutsCh:
+				// Handling of timeouts is priorities over messages
+				switch tm.s {
+				case propose:
+					t.OnTimeoutPropose(tm.h, tm.r)
+				case prevote:
+					t.OnTimeoutPrevote(tm.h, tm.r)
+				case precommit:
+					t.OnTimeoutPrecommit(tm.h, tm.r)
+				}
+
+				i := slices.Index(t.scheduledTms, tm)
+				t.scheduledTms = slices.Delete(t.scheduledTms, i, i+1)
+			case p := <-t.proposalsCh:
+				t.handleProposal(p)
+			case p := <-t.listeners.ProposalListener.Listen():
+				t.handleProposal(p)
+			case p := <-t.prevotesCh:
+				t.handlePrevote(p)
+			case p := <-t.listeners.PrevoteListener.Listen():
+				t.handlePrevote(p)
+			case p := <-t.precommitsCh:
+				t.handlePrecommit(p)
+			case p := <-t.listeners.PrecommitListener.Listen():
+				t.handlePrecommit(p)
+			}
+		}
+	}()
+}
+
+func (t *Tendermint[V, H, A]) Stop() {
+	close(t.quit)
+	for _, tm := range t.scheduledTms {
+		tm.Stop()
+	}
+	t.wg.Wait()
+}
+
+func (t *Tendermint[V, H, A]) startRound(r round) {
+	if r != 0 && r <= t.state.r {
+		return
+	}
+
+	t.state.r = r
+	t.state.s = propose
+
+	t.state.timeoutPrevoteScheduled = false
+	t.state.lockedValueAndOrValidValueSet = false
+	t.state.timeoutPrecommitScheduled = false
+
+	if p := t.validators.Proposer(t.state.h, r); p == t.nodeAddr {
+		var proposalValue *V
+		if t.state.validValue != nil {
+			proposalValue = t.state.validValue
+		} else {
+			v := t.application.Value()
+			proposalValue = &v
+		}
+		proposalMessage := Proposal[V, H, A]{
+			H:          t.state.h,
+			R:          r,
+			ValidRound: t.state.validRound,
+			Value:      proposalValue,
+			Sender:     t.nodeAddr,
+		}
+
+		t.messages.addProposal(proposalMessage)
+		t.broadcasters.ProposalBroadcaster.Broadcast(proposalMessage)
+	} else {
+		t.scheduleTimeout(t.timeoutPropose(r), propose, t.state.h, t.state.r)
+	}
+
+	go t.processFutureMessages(t.state.h, t.state.r)
```

Comment:
Is there any reason why we concurrency here? The algorithm is mostly sequential, so I think we can avoid this mutex and 3 channels for future by moving the events out of `futureMessages`, maybe to a field instead of pushing them to 3 channels. That would avoid a mutex, a concurrent goroutine and context switching over 3 more channels.

Reply:
Maybe it is possible, i didn't try to think of a way to avoid it but couldn't come up with one. The main reason for it was that we need to process the future messages as soon as possible and the easiest way to do that way through channels also while processing the future messages we cannot hold the processing of the new messages which arrive.